src/decision/justification_strategy.cpp

```diff
@@ -0,0 +1,638 @@
+          // beginning of the next call to getNext above.
+          d_lastDecisionLit = next.first;
+          // record that we made a decision
+          d_currStatusDec = true;
+          return lastChildVal == SAT_VALUE_FALSE ? ~nsl : nsl;
+        }
+        else
+        {
+          // NOTE: it may be the case that we have yet to justify this node,
+          // as indicated by the return of lookupValue. We may have a value
+          // assigned to next.first by the SAT solver, but we ignore it here.
+          // (2) unprocessed non-atom, push to the stack
+          d_stack.pushToStack(next.first, next.second);
+          d_stats.d_maxStackSize.maxAssign(d_stack.size());
+          // we have yet to process children for the next node, so lastChildVal
+          // remains set to SAT_VALUE_UNKNOWN.
+        }
+      }
+      else
+      {
+        Trace("jh-debug") << "in main loop, " << next.first << " has value "
+                          << lastChildVal << std::endl;
+      }
+      // (3) otherwise, next already has a value lastChildVal which will be
+      // processed in the next iteration of the loop.
+    }
+  } while (d_stack.hasCurrentAssertion());
+  // we exhausted all assertions
+  Trace("jh-process") << "...exhausted all assertions" << std::endl;
+  stopSearch = true;
+  return undefSatLiteral;
+}
+
+JustifyNode JustificationStrategy::getNextJustifyNode(
+    JustifyInfo* ji, prop::SatValue& lastChildVal)
+{
+  // get the node we are trying to justify and its desired value
+  JustifyNode jc = ji->getNode();
+  Assert(!jc.first.isNull());
+  Assert(jc.second != SAT_VALUE_UNKNOWN);
+  // extract the non-negated formula we are trying to justify
+  bool currPol = jc.first.getKind() != NOT;
+  TNode curr = currPol ? jc.first : jc.first[0];
+  Kind ck = curr.getKind();
+  // the current node should be a non-theory literal and not have double
+  // negation, due to our invariants of what is pushed onto the stack
+  Assert(!isTheoryAtom(curr));
+  Assert(ck != NOT);
+  // get the next child index to process
+  size_t i = ji->getNextChildIndex();
+  Trace("jh-debug") << "getNextJustifyNode " << curr << " / " << currPol
+                    << ", index = " << i
+                    << ", last child value = " << lastChildVal << std::endl;
+  // NOTE: if i>0, we just computed the value of the (i-1)^th child
+  // i.e. i == 0 || lastChildVal != SAT_VALUE_UNKNOWN,
+  // however this does not hold when backtracking has occurred.
+  // if i=0, we shouldn't have a last child value
+  Assert(i > 0 || lastChildVal == SAT_VALUE_UNKNOWN)
+      << "in getNextJustifyNode, value given for non-existent last child";
+  // In the following, we determine if we have a value and set value if so.
+  // If not, we set desiredValue for the value of the next value to justify.
+  // One of these two values should always be set below.
+  SatValue value = SAT_VALUE_UNKNOWN;
+  SatValue desiredVal = SAT_VALUE_UNKNOWN;
```

Comment:
This comment does not really explain what the actual difference between `value` and `desiredVal` is.